---
title: "vaRHC"
header-includes:
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
author: "Elisabet Munté Roca <emunte@idibell.cat>"
date: "`r Sys.Date()`"
package: "`r pkgload::pkg_version(path='.')"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{vaRHC}
  %\VignetteEngine{knitr::knitr}
  %\usepackage[utf8]{inputenc}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(dplyr)
```
## Contents

  1. [Introduction](#intro)
  2. [Installation](#install)
  3. [Main functions](#functions)<br><br>
    3.1 [vaR()](#func1) <br>
    3.2 [vaRbatch()](#func2)<br>
  4. [Legal advice and privacy policy for users](#legal)
  5. [Supplementary Tables](#Supplement)
  6. [Session Info](#session)
  


## 1. Introduction {#intro}

Variant classification is a manual complex long process that combines information of distinct nature. An accurate classification is necessary to ensure a proper genetic counselling and personalized risk estimation. 

In 2015, the American College of Molecular Genetics and Genomics (ACMG) together with the Association of Molecular Pathologists published generic guidelines to standardize and provide an objective framework to evaluate variant pathogenicity in Mendelian disease.
Later, specific guidelines have been published for some genes by collaborative groups. 

Moreover, it has been demonstrated that the combination of criteria in ACMG/AMP guidelines is compatible with a quantitative Bayesian formulation(Tavtigian et al., 2018) and a naturally scaled point system has been further abstracted (Tavtigian et al., 2020). Additionally, CanVIG-UK consensus recommendations has proposed some limitations to overlapping criterion combination to avoid double counting of evidence (Garrett et al., 2021).  

In the last 5 years, some programs have been developed with the aim to semi-automatize the process of variant classification. Most tools are based on ACMG-AMP general rules but others focus on a set of genes. 

`vaRHC` has been developed to automate as much as possible the process of variant classification in hereditary cancer (HC). The aim is to streamline the work of biologists and avoid possible manual error following gene-specific guidelines for *APC*, *ATM*, *CDH1*, *CHEK2*, *MLH1*, *MSH2*, *MSH6*, *PALB2*, *PMS2*, *PTEN* and *TP53* and the updated general ACMG rules for the remaining cancer susceptibility genes. The final classification is obtained according to  Tatigian's natural scoring Bayesian-based metastructure (Tavtigian et al., 2020) but also considering the  CanVIG-UK incompatibilities proposal. vaRHC gives the oportunity to export the output in a .xlsx file as a user-friendly way to examine and store the results allowing non-bioinformatic users to work with them, and even modify the file adding their considerations or information regarding the non-automatable criteria. 

The current version of the package  works for single substitutions, deletions and insertions up to 25 bp, intronic variants and 5’ or 3’-UTR variants.

## 2. Installation {#install}
vaRHC can be downloaded from CRAN:
```{r, eval=FALSE}
install.packages("vaRHC")
```
vaRHC can also be downloaded from GitHub using the remotes package:
```{r, eval=FALSE}
if(!require("remotes", quietly = TRUE)) install.packages('remotes') ## Only the first time
library(remotes)
devtools::install_github("emunte/vaRHC")
```


Next, it has to be loaded into the workspace:
```{r}
library("vaRHC")
```

## 3. Main functions {#functions}
The package consists of two main functions ***vaR()***  and ***vaRbatch()***.

### 3.1 vaR() {#func1}

#### 3.1.1 Input
##### 3.1.1.1 Parameters
Using a **gene** and a **variant** name (in coding DNA nomenclature) and if possible **NM** from RefSeq as input variables,  ***vaR()*** gathers relevant information from different sources.  Gene-specific ACMG/AMP guidelines are then applied in order to calculate whether the variant of interest meets different criteria. The output also provides an explanation of the reason for applying or rejecting each criterion. Additionally, it returns a final classification of the variant using Tavtigian Bayesian metastructure and also considering most of CanVIG-UK recommendations.


The following example shows how to call the **vaR()** function for the variant  c.1137+1G>A in *CDH1* gene.

```{r, eval=FALSE, cache=TRUE}
eg.gene <- "CDH1"
eg.variant <- "c.1137+1G>A"


var.information <- vaR(gene = eg.gene, variant = eg.variant)

```

##### 3.1.1.2 Optional parameters
Optional parameters of *vaR()* are described below.

##### 3.1.1.2.1 NM and CCDS
- ***NM***: Accession number of the transcrit and mRNA from RefSeq. By default is NULL and vaRHC will consider the ones detailed above. Be careful if you use a different NM because the program has not been validated for it. If you provide a different NM, CCDS must also be provided.The transcript accession number is retrieved from RefSeq. By default, this function parameter is set to NULL. vaRHC will consider the NMs listed in the Supplementary Table 1. If a different NM is used than the ones described in the table, it must be noted that the program has not been validated yet for it. 

- ***CCDS***: Consensus CDS ID retrieved from NCBI CCDS database (which can be found in the following link https://www.ncbi.nlm.nih.gov/projects/CCDS/CcdsBrowse.cgi). By default is set to NULL. vaRHC will consider the IDs detailed in the table above. 


##### 3.1.1.2.2 gene.specific.df
**gene.specific.df**: By default the parameter is set to NULL as it considers the gene-specific cutoffs, frequencies and additional parameters queried in the IDIBELL database as described in the paper *vaRHC: an R package for semi-automation of variant classification in hereditary cancer genes according to ACMG/AMP and gene-specific ClinGen guidelines [Manuscript submitted for publication]*.
This table contains gene-specific information for applying the following criteria: BA1, BS1, BS2, PM2, predictors cut-off and BP7. If the user wants to modify any of these criteria or specify another additional gene, a table containing all these criteria can be loaded to the R environment, and set as the gene.specific.df parameter. 
A template of how the table needs to be can be downloaded from GitHub (https://github.com/emunte/vaRHC/blob/main/data/gene_specific.txt) or it can be found at the package documentation. Column names cannot be modified. See below the default table and the explanation of each column:

\newpage
\blandscape
Landscape
```{r, echo=FALSE}
 gene.specific.df <- vaRHC:::connectionDB("SELECT * from  gene_specific;") %>% as.data.frame()
    knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53")) %>% dplyr::select(1:10))
     knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53")) %>% dplyr::select(11:20))
    knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53"))%>% dplyr::select(21:30))
       knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53"))%>% dplyr::select(31:40))
       knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53"))%>% dplyr::select(41:50))
       knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53"))%>% dplyr::select(51:58))
       knitr::kable(gene.specific.df %>% filter(genename %in% c("general", "ATM", "CDH1", "CHEK2", "MLH1", "MSH2", "MSH6", "PMS2", "PTEN", "TP53"))%>% dplyr::select(59:67))
```
\elandscape

- **BA1**: *float number*. Cut-off used to assign BA1 criterion.<br>
- **BS1**: *float number*. Cut-off used to assign BS1 criterion.<br>
- **IC**: *float number*. Interval of confidence used to assign BA1 and BS1 criterion<br>
- **alleles**: *int number*. Minimum number of alleles needed in the subpopulation (it will only be considered when IC is not specified).<br>
- **BS2**: *int number*. Minimum number of individuals that need to carry the variant to assign BS2.<br>
- **BS2_sup**: *int number*. Minimum number of individuals that need to carry the variant to assign BS2_sup.<br>
- **BS2_db**: *character* Db to be queried for BS2 criterion. Please introduce one of the following options *FLOSSIES*, *GNOMAD_non_cancer*, *GNOMAD_non_neuro* or *NA*. .<br>
- **status**: *character* Type of zigosity asked for BS2 criterion. Please introduce one of the following options *homo_healthy*, *hete_healthy*, *NA*. <br>
- **PM2**: float number. Cut-off used to assign PM2 criterion.<br>
- **PM2_sup**: float number. Cut-off used to assign PM2_supporting criterion.<br>
- **predictors section**:<br>
  Predictors shown are: nucleotide conservation(**phylop, phastcons, gerp**), protein level (**revel, VEST4, provean, bayesDel_noAF, agvgd, polyphen, MAPP, prior_utah_prot**) and splicing predcitors (**spliceai, trap**). See references in the tool paper. <br>
  For each predictor there are 4 columns: <br>
  + **op_predictor_ben**: *character* Operator to use for benign cut-off. The possible options to introduce are *<*, *>*, *=<*, *=>* or *!=*. 
  + **op_predictor_pat** : *character* Operator to use for pathogenic cut-off. The possible options to introduce are *<*, *>*, *=<*, *=>* or *!=*. 
  + **predictor_ben** : *float number*. Benign cut-off to use with this predictor
  + **predictor_pat** : *float number*. Pathogenic cut-off to use with this predictor<br>
  
> Be careful, changing a predictor cut-off will only be considered when the guidelines specify to use that predictor.

- **BP7_splicing**: *character* Wheter the BP7 criteria is dependent on splicing prediction not being altered or not. Possible options are *dependent*, *indpendent*.  
```{r, eval=FALSE}
eg.gene <- "CDH1"
eg.variant <- "c.1137+1G>A"
data("gene_specific")
eg.gene.specific <- gene_specific

var.information <- vaR(gene = eg.gene, variant = eg.variant, gene.specific.df = eg.gene.specific)
```

##### 3.1.1.2.3 Connecting to webpages using javascript
Insight database needs to be queried using RSelenium package. However, we have detected that some Institutes have unabled the possibility to connect to a remote server. The ****remotes*** parameter allows to decide if the user wants to allow this capability (TRUE) or not (FALSE). If it is set to FALSE the user will not collect information from insight database. 
The ***browser*** parameter is used to set which browser to start Rselenium server. By default is "firefox" (the recommended). If you do not have firefox installed try either "chrome" or "phantomjs" (but they have not been tested for this package).

##### 3.1.1.2.4 SpliceAI related parameters
- ***spliceai.program*** : *Logical*. By default is FALSE , assuming that SpliceAI program is not installed in your computer. If this parameter is FALSE, the program will only classify substitutions and simple deletion variants considering a spliceAI distance of 1000 and will show masked results for *ATM*, *BRCA1*, *BRCA2*, MMR genes, *TP53* and *PTEN*. If you want to classify other variants please install SpliceAI (https://pypi.org/project/spliceai/) and set the parameter to TRUE.

- ***spliceai.genome***: It can be either "hg19" or "hg38". By default "hg19".
- ***spliceai.reference*** : Path to the Reference genome hg19 or hg38 fasta file. Hg19 file can be downloaded from http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz . By default is NULL and it will only be taken into account if spliceai.program is set to TRUE.

- ***spliceai.annotation*** : Path to gene annotation file. By default is null and it uses UCSC ncbiRefSeq table. This page sometimes is out of service. Then the file must be provided (see and example in package docs folder: "../docs/gencode.v38lift37.annotation.txt")

- ***spliceai.distance*** :  *Integer*. Maximum distance between the variant and gained/lost splice site (default: 1000)

- ***spliceai.masked*** : Mask scores representing annotated acceptor/donor gain and unannotated acceptor/donor loss (default: 1, that is masked)

- ***spliceai.10k*** : *Logical*. By default is FALSE. If TRUE performs SpliceAI.10k calculation. Be aware that it will increase dramatically the time needed per variant.

##### 3.1.1.2.5 Provean related parameters
- ***provean***: *Logical*. By default is FALSE and it is assumed that provean program is not installed in your computer. Set to TRUE if you want to compute provean locally.
- ***provean.sh***: Path to the provean.sh file. It will only be considered when provean is set to TRUE.

##### 3.1.1.2.6 Report related parameters

- ***excel.results*** : *Logical*. By default is FALSE and the excel file will not be created. If TRUE and excel file will be saved.
- ***path.original.file***: If excel.results param is set to TRUE, path.original.file must contain the path to the excel template. By default is the template located in the package docs folder.


##### 3.1.1.2.7 Output dir
- ***output.dir***:  By default is NULL and the output will be saved in the working directory. If you want to save the output in another folder please enter here the path. 

####  3.1.2 Output
The result is stored in a list but it can also be exported in a xlsx file.
The list has to dimensions.

##### 3.1.2.1 vaRinfo
Where all the information retrieved is stored. The names of the list elements are the following.

```{r, include=FALSE, cache=TRUE}
eg.gene <- "CDH1"
eg.variant <- "c.1137+1G>A"

#var.all <- vaR(assembly = eg.assembly, gene = eg.gene, variant = eg.variant, gene.specific.df = NULL)
var.information <- vaR(variant = eg.variant, gene= eg.gene, gene.specific.df = NULL)

```
```{r}
names(var.information$vaRinfo)

```

- **Variant.Info**: a data.frame containing a summary of variant location, different nomenclatures and variant consequences.
```{r, echo=F}
var.info <- var.information$vaRinfo$Variant.Info %>% as.data.frame()
str(var.info)
```
- **Variant.Info.other**: only for *TP53* and *CDK2NA* other transcripts.
- **variant.correction**: output obtained by quering Mutalyzer.
- **gene.specific.info**: variant gene specificities used
- **gnomAD**: a list containing gnomAD v2.1.1 variant nomeclature, coverage and alleles information for non_cancer and non_neuro datasets. See below examples of how to obtain the gnomAD data
```{r}
#nomenclature
var.information$vaRinfo$gnomAD$nomenclature
#coverage
var.information$vaRinfo$gnomAD$coverage
#information from exomes non cancer separated by subpopulations
knitr::kable(var.information$vaRinfo$gnomAD$info$exomes$non.cancer$subpopulations)
#information from exomes + genomes non neuro overall frequency
knitr::kable(var.information$vaRinfo$gnomAD$info$exomes.genomes$non.neuro$overall)
```
- **flossies.db**: a dataframe containing information obtained from FLOSSIES database.
- **clinVar**: a list containing information of the variant of interest and if it is a missense also information from other missense variants located at the same codon. 
- **predictors**: a datafrane with all the predictors information stored. Only the predictors where use column is yes will be considered to calculate criteria.
```{r, echo=FALSE}
knitr::kable(var.information$vaRinfo$predictors$predictor.table)
```
- **codon.stop**: a list with the following elements:
  + **variant.exon**: contains the exon where the variants is located and its coordinates.
  
```{r, echo=FALSE}
  knitr::kable(var.information$vaRinfo$codon.stop$variant.exon)
```
  + **premature.ter.codon**: for frameshifts and nonsense variants it contains the exon where the stop codon is produced and its coordinates.
  + **length.transcript**: total number of coding nucleotides that the transcript has
  + **porc.prot**: percentatge of protein conserved
  + **canonical.skip.pred**: only for canonical splice variants. It contains information from the predicted skipping variant, its consequence and percentatge of protein conserved. 
```{r, echo=FALSE}
  knitr::kable(var.information$vaRinfo$codon.stop$canonical.skip.pred)
```
  + **exons**: exon coordinates according to LRG transcript. In genomic and coding dna nomenclature.
```{r, echo=FALSE}
  knitr::kable(var.information$vaRinfo$codon.stop$exons)
```
- **second.met**: only for start codon variants. It returns second metionine position and clinVar variants between first and second metionine.
- **insight.info**: information from INSIGHT database classifications and MMR Integrative Evaluation (http://www.insight-database.org/classifications/ and http://www.insight-database.org/classifications/mmr_integrative_eval.html )
- **functional.assays**: a list containing information from functional assays. Further details are explained in the tool article. 
- **cancer.hotspots**: information obtained from Cancer Hotspots (http://www.insight-database.org/classifications/mmr_integrative_eval.html)

#### 3.1.2.2 vaRclass
All the assigned and denied criteria are stored in this list as well as final  classification. 

The list has the following items:
```{r}
names(var.information$vaRclass)
```
The final.classification element contains:<br>

  + **final.class**: variant's final classification
  + **criteria.assingned**: a vector containing all the criteria that are assigned to the variant
  + **sum.criteria**: Tavtigian's Bayesian score
  + **discrep.reason**: wheter there are discrepances or not and the reason.
  
  
The final.criteria element contains: 
 + *Criteria.res*: a matrix with all the criteria assigned(1), denied (0), not applicable (NA) or not calculated (NC).
 + *Criterion name + .message*: reasoning for the criterion being assigned or denied.
```{r echo=FALSE}
knitr::kable(var.information$vaRclass$final.criteria$criteria.res)
var.information$vaRclass$final.criteria[2:length(var.information$vaRclass$final.criteria)]
```

3.1.2.3 Excel file
The excel file has the following sheets:<br>

+ **Classification summary**: variant summary and criteria assigned.
+ **Evidence**: justification of every evidence given to a variant. 
+ **Frequency control**: population dta from gnomAD v2.1.1 non_cancer, non_neuro, FLOSSIES database and cancer hotspots
+ **ClinVar**: variant information from ClinVar.
+ **ClinVar variants**: List of variants in ClinVar at the same codon(only for missense variants).
+ **ClinVar + variant name**: One sheet per variant listed in the previous sheet containing all the information provided by ClinVar.
+ **Predictors**: Variant scores for the predictors listed and its cut-offs.
+ **NMD**: Information for loss of function variants due to frameshift, nonsense or canonical splice site alterations. 
+ **Start Codon**: for start codon variants: list of ClinVar variants upstream the second in-frame metionine
+ **Bibliography vaRCH**: a string containing all possible variant nomenclature to put in a search engine such as google. It also returns the 30 first articles listed in google scholar. Be careful because Google scholar has a limited of searches per day. When this searches are overpasses an Error 429 is returned.  



### 3.1 vaRbatch() {#func2}
***vaRbatch()***  allows to perform vaR function in batch.

#### 3.1.1 Input
vaRbatch has two possible inputs:
- a data frame containing gene, variant and NM (not necessary but it is recommended, otherwise the function will consider NM detailed in supplementary table 1). 
- a small variant call format (vcf or compressed vcf). The file will be annotated using Mutalyzer v3 API. Vcfs resulting from WES or WGS will take too long (although they work as well). Assembly parameter must be then specified


```{r eval=FALSE}
#Example 1
output.dir <- "/path/to/output_dir"
#Example 1
data("ex_vaRbatch")
ex_vaRbatch[] <- lapply(ex_vaRbatch, as.character) #convert to character
all <- vaRbatch( all.variants = ex_vaRbatch, spliceai.program = TRUE,  splieai.reference= "dir to /hg19.fa", output.dir = excel.results = TRUE)

#Example2
eg.variants <- data.frame(gene=c("ATM", "MSH6", "BRCA1"), variants = c("c.8420A>T", "c.1559G>A", "c.211A>G"))
batch.results <- vaRbatch( all.variants = eg.variants, spliceai.program =FALSE, output.dir = ouput.dir, print.data.frame = FALSE excel.results = TRUE)

#Example 3
vcf.file <- "path/to/vcf.vcf"
results <- vaRbatch(all.variants= vcf.file, assembly="hg19", spliceai.program = TRUE, spliceai.reference = "/path/to/vcf.vcf", remote=TRUE, output.dir = output.dir, excel.results = FALSE, print.data.frame = TRUE)
```

#### 3.1.2 Output
It returns a list containing vaR() output for all variants. The user can choose to print a data frame (***print.data.frame***) to store classification results in a csv file. The function also returns a log file, with the detailed time execution and any error that may occur explained. The log file is stored in a log folder in the specified output.dir directory.


### 4. Legal advice and privacy policy for users {#legal}
#### Legal advice
##### Terms and conditions 
Access and navigation through the web application ascribe the User condition and imply the full and unreserved acceptance of each of the provisions included in this Legal Warning and Privacy Policy.  
Also, for any doubt the user has regarding using the software, he can contact emunte@idibell.cat. 

##### User's rights and obligations 
The user has the right to browse the website, observe the rules established in its notices and policies at all times, and the right to have your information processed with respect and maximum privacy, which is why we have provided adequate Privacy and Data Protection Policy. <br>

The user must always respect the terms and conditions established in this legal notice. Expressly, the user states he will use the portal diligently and assume any liability that may arise from the breach of rules. <bR>

Users will be obliged to make reasonable use of services or content, under the principle of good faith and respect for the legality in force, morality, public order, good customs, third-party rights, or IDIBELL itself, all according to the possibilities and objectives for which they are conceived. <br>

IDIBELL does not assume direct or indirect responsibilities for emerging damage or enduring profit resulting from users or third parties' misuse of services or content. <br>

The user undertakes to indemnify and hold harmless the website for any damage, prejudice, penalty, fine, penalty, or compensation that may have to have the website. 

 
##### Confidentiality rules
Your data may be collected when the user browses the software, and, in this case, IDIBELL will respect the provisions of the Privacy Policy. <br>

The application can automatically detect the user's IP address and domain name. An IP address is a number automatically assigned to a computer when it connects to the Internet. All this information is registered in a server activity file duly registered that allows the subsequent processing of the data to obtain only statistical measurements that allow knowing the number of page impressions, the number of visits made to the web servers, the order of visits, the access point, and to assess other Website' aspects. 

##### Software of the web and its design 
IDIBELL is the owner of this software because it has been developed by its hired staff. 

##### Responsibility

IDIBELL does not guarantee the continuous and permanent availability of the services, and IDIBELL has not any responsibility for possible damages caused as a result of the lack of availability of the service due to force majeure or errors in the telematic networks of transfer data, alien to their will, or disconnections made for improvement or maintenance of equipment and computer systems.<br>

Also, not responsible for possible omissions, loss of data, settings, improper access, or breach of confidentiality that have their origin in technical problems, communications, or human failures, caused by third parties or not attributable to the software. Nor will it be liable for damages caused by computer attacks or caused by viruses that affect computer programs, communications systems, or equipment used by the website but manufactured or provided by a third party. <br>

IDIBELL will not be able to be responsible for any service suspension that is determined by a internal decision. <br> 

IDIBELL will be able to its discretion, deny, withdraw, suspend and/or block at any time, and without warning, access to information and services to users who breach the present rules except where the Act expressly imposes the opposite, and exclusively with the measure and extent to which it assesses, no liability is guaranteed or assumed for damages caused by software, data and portal services. <br>

##### Applicable law and jurisdiction

IDIBELL also reserved the right to file civil or criminal actions it deems appropriate for the improper use of its website and content or breaking these conditions.<br>

Current Spanish regulations will govern the relationships between IDIBELL and User, and any dispute will be submitted to the Court and Tribunals of the city of Barcelona. <br>


#### Privacy Policy 

In compliance with the obligations established in Organic Law 3/2018, of December 5, Protection of Personal Data and guarantee of digital rights and the General Regulation of Data Protection (RGPD), which is the Regulation (EU) 2016/679 of the Parliament and of the Council, of April 27, 2016, it is reported that the person in charge of the treatment of your personal data is Institut d'Investigació Biomètrica de Bellvitge (IDIBELL)CIF G58863317, with address Av/ Gran via de L'Hospitalet, nº 199-203, 08908, Hospitalet de Llobregat, his personal data included in the software (access IP), in order to manage its associated services.  <br>

The application only uses its session cookies for technical purposes. <br>

You will be able to request more information on the processing of data performed and exercise access rights, rectification, deletion, opposition, portability, and limitation of the process by sending a request through the email address dataprotection@idibell.cat. You will also be able to contact the data protection delegate of the IDIBELL via dpd@ticsalutsocial.cat. <br>  

Furthermore, if you consider that you have not obtained satisfaction in the exercise of your rights, you will be able to file a complaint before the Catalan Data Protection Agency, and you will be able to find more information on the processing of your data from the IDIBELL to the following Privacy Policy. <br>



## 5. Supplementary Tables {#Supplement}

### ST1
```{r echo=FALSE}
all.transcript <- vaRHC:::connectionDB("SELECT ensembltranscriptID, namegene, NM, NC, CCDS from transcript WHERE maintranscript = 'T';") %>% as.data.frame() %>% dplyr::arrange(namegene)
knitr::kable(all.transcript)
```

### 6. Session info {#session}
```{r}
sessionInfo()
```
